### PR TITLE
fix simple tree shortId bug

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR fixes a bug in the simple tree's shorId api, in the case where the user provides an invalid stable id in the identifier field.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/flex-tree/nodeKeys.ts [hunk 1]
```diff
@@ -29,6 +29,12 @@ export interface NodeKeys {
 	 * @param key - the key to convert
 	 */
 	localize(key: StableNodeKey): LocalNodeKey;
+	/**
+	 * Attempts to recompress a {@link StableNodeKey} back into its {@link LocalNodeKey} form.
+	 * @param key - The key to convert.
+	 * @returns The {@link LocalNodeKey} associated with `key` or undefined if the UUID was not generated by any session known to this compressor.
+	 */
+	tryRecompress(key: StableNodeKey): LocalNodeKey | undefined;
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/flex-tree/nodeKeys.ts:37]: We should call this `tryLocalize` since it is the the "try-variant" of `localize`.

Also, the parameter `key` should be typed `string` rather than `StableNodeKey`. Here is the rationale. According to this interface, the _only_ way for a user to get a `LocalNodeKey` (initially) is to generate one with `generate()`. That `LocalNodeKey` is _guaranteed_ to succeed when passed to `stabilize()`, and the resulting `StableNodeKey` is guaranteed to succeed when passed to `localize()`. These functions are all guaranteed to succeed so long as the user "obeys the types" - i.e. _they don't cast_. We got into trouble and created a bug because we did a cast. So, let's continue to structure this interface with the same assumption - _no casting allowed_ - and let's find a way to remove the cast you are doing at the call site. With that in mind, a function `tryRecompress(key: StableNodeKey): LocalNodeKey | undefined` doesn't make sense, because a true `StableNodeKey` will never return `undefined`. But a function like `tryRecompress(key: string)` does make sense - it's saying "hey, I have some string and I'm not sure if it is a StableNodeKey or not, but if it is, can you get me the `LocalNodeKey` for it?"

Hopefully that makes sense.

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 2]
```diff
@@ -33,6 +33,13 @@ export interface NodeKeyManager {
 	 * Convert the given {@link LocalNodeKey} into its {@link StableNodeKey} form.
 	 */
 	stabilizeNodeKey(key: LocalNodeKey): StableNodeKey;
+
+	/**
+	 * Attempts to recompress a StableNodeKey.
+	 * @param key - The UUID that is attempted to recompress.
+	 * @returns The `LocalNodeKey` associated with `key` or undefined if the UUID was not generated by any session known to this compressor.
+	 */
+	tryRecompress(key: StableNodeKey): LocalNodeKey | undefined;
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:42]:
```suggestion
	tryLocalizeNodeKey(key: string): LocalNodeKey | undefined;
```

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 3]
```diff
@@ -99,6 +116,11 @@ class MockNodeKeyManager implements NodeKeyManager {
 		return brand(this.createMockStableId(extractFromOpaque(key)));
 	}
 
+	public tryRecompress(key: StableNodeKey): LocalNodeKey | undefined {
+		const localNodeKey = this.localizeNodeKey(key);
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:120]: Typically it's best to write "non-try" functions in terms of "try" functions, rather than the other way around, because it puts all of the interesting logic in the "try" function and makes the "non-try" function trivial to write as a wrapper around the "try" function. The "try" function here should accept a `string`, and if the string is a valid mock UUID* it succeeds otherwise it returns undefined. The "non-try" function (`localizeNodeKey`) should just call this "try" function and fail if it returns undefined.

*A valid mock UUID is one like `a110ca7e-add1-4000-8000-XXXXXXXXXXXX`, where XXXXXXXXXXXX is less than `this.count`.

File: packages/dds/tree/src/simple-tree/schemaFactory.ts [hunk 4]
```diff
@@ -611,7 +611,7 @@ export class SchemaFactory<
 	 * Make a field of type identifier instead of the default which is required.
 	 */
 	public get identifier(): FieldSchema<FieldKind.Identifier> {
-		return createFieldSchema(FieldKind.Identifier, this.string);
+		return createFieldSchema(FieldKind.Identifier, [this.string, this.number]);
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/schemaFactory.ts:614]: Why this change?

File: packages/dds/tree/src/feature-libraries/flex-tree/nodeKeys.ts (hunk 1 above)
Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/flex-tree/nodeKeys.ts:37]: Updated function name and type

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts (hunk 2 above)
Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts]: Updated function name and type

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts (hunk 3 above)
Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts]: Updated non-try function to call the try function and fail when it returns undefined

File: packages/dds/tree/src/simple-tree/schemaFactory.ts (hunk 4 above)
Comment (daesunp) [reply packages/dds/tree/src/simple-tree/schemaFactory.ts]: Reverted change, and updated TreeApi shortID return types

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 5]
```diff
@@ -92,13 +109,26 @@ class MockNodeKeyManager implements NodeKeyManager {
 	}
 
 	public localizeNodeKey(key: StableNodeKey): LocalNodeKey {
-		return brand(Number.parseInt(key.substring(30), 16) as SessionSpaceCompressedId);
+		const localNodeKey = this.tryLocalizeNodeKey(key);
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:112]: You can use `?? fail("Key is not compressible")` to do this whole method in one line - it's the pattern preferred by much of the ST codebase.

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 6]
```diff
@@ -92,13 +109,26 @@ class MockNodeKeyManager implements NodeKeyManager {
...
+		return localNodeKey;
 	}
 
 	public stabilizeNodeKey(key: LocalNodeKey): StableNodeKey {
 		return brand(this.createMockStableId(extractFromOpaque(key)));
 	}
 
+	public tryLocalizeNodeKey(key: string): LocalNodeKey | undefined {
+		const localNodeKey = Number.parseInt(key.substring(24), 16);
+		if (Number.isNaN(localNodeKey)) {
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:123]: You can use `isStableId` to validate that `key` is a StableNodeKey, and if so you can assume that the parseInt will succeed - no need for the isNaN check.

Comment (noencke) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:123]: Or better, I would consider making an `isStableNodeKey()` type guard (which just calls `isStableId` inside) for maximum type safety.

File: packages/dds/tree/src/simple-tree/treeApi.ts [hunk 7]
```diff
@@ -96,9 +96,9 @@ export interface TreeNodeApi {
 
 	/**
 	 * If the given node has an identifier specified by a field of kind "identifier" then this returns the compressed form of that identifier.
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/treeApi.ts:98]: This function now returns the "most compressed" form of the ID, if there is one, so I think we should reword it a bit. Maybe something like:

Returns the {@link FieldKind.Identifier | identifier} of the given node in the most compressed form possible.
@remarks
If the node's identifier is a valid {@link StableNodeKey}, then this will return a unique process-local integer corresponding to that identifier.
If the node's identifier is any other string, then this will return that string.
If the node has no identifier (that is, it has no field of an {@link FieldKind.Identifier | identifier} field kind), then this returns `undefined`.

File: packages/dds/tree/src/simple-tree/treeApi.ts [hunk 8]
```diff
@@ -189,16 +189,18 @@ export const treeNodeApi: TreeNodeApi = {
...
 		const flexNode = getFlexNode(node);
 		for (const field of flexNode.boxedIterator()) {
 			if (field.schema.kind === FieldKinds.identifier) {
 				const identifier = field.boxedAt(0);
 				assert(identifier !== undefined, 0x927 /* The identifier must exist */);
-
-				return extractFromOpaque(
-					identifier.context.nodeKeys.localize(identifier.value as StableNodeKey),
+				const localNodeKey = identifier.context.nodeKeys.tryLocalize(
+					identifier.value as StableNodeKey,
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/treeApi.ts:199]: Remove this cast - cast to string instead. We _know_ that value is a string, so that's a fine cast. We do _not_ know that it is a `StableNodeKey`, so that's _not_ a fine cast!

You're also casting to string below - you can just cast once above: `const identifier = field.boxedAt(0) as string`.

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 9]
```diff
@@ -68,6 +75,16 @@ export function createNodeKeyManager(idCompressor?: IIdCompressor | undefined):
 				assertIsStableId(idCompressor.decompress(extractFromOpaque(key))),
 			);
 		},
+		tryLocalizeNodeKey: (key: string) => {
+			assert(
+				idCompressor !== undefined,
+				0x6e9 /* Runtime IdCompressor must be available to convert node keys */,
+			);
+			const compressedKey = idCompressor.tryRecompress(key as StableNodeKey);
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:83]: `key as StableNodeKey`

no cast! cast bad! :D

Use `isStableId()`. IdCompressor only requires that the input is a `StableId`, it doesn't care about `StableNodeKey`.

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts (hunk 6 above)
Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts]: Added `isStableNodeKey` type guard function to `nodeKeyManager` and `nodeKeys` and removed `isNaN` check

File: packages/dds/tree/src/simple-tree/treeApi.ts (hunk 7 above)
Comment (daesunp) [reply packages/dds/tree/src/simple-tree/treeApi.ts]: Updated docstring to suggested change

File: packages/dds/tree/src/simple-tree/treeApi.ts (hunk 8 above)
Comment (daesunp) [reply packages/dds/tree/src/simple-tree/treeApi.ts]: I couldn't access `isStableId` from `treeApi.ts`, so I added a type guard function `isStableNodeKey` to `nodeKeys` and `nodeKeyManager` so that it could be used here

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts (hunk 9 above)
Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts]: Updated to use `isStableNodeKey` and removed bad cast

File: packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts [hunk 10]
```diff
@@ -33,6 +34,18 @@ export interface NodeKeyManager {
...
+	 * Attempts to recompress a StableNodeKey.
+	 * @param key - The UUID that is attempted to recompress.
+	 * @returns The `LocalNodeKey` associated with `key` or undefined if the UUID was not generated by any session known to this compressor.
+	 */
+	tryLocalizeNodeKey(key: string): LocalNodeKey | undefined;
+
+	/**
+	 * type guard to check if a key is a {@link StableNodeKey}
+	 */
+	isStableNodeKey(key: string): key is StableNodeKey;
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts:48]: I see two options for this function:
1. It is a free function (not an interface member) and all it does is call `isStableId()`. You want it to be a free function, not an interface member, because otherwise every implementation of the interface has to implement it the same way - it is needless duplication.
2. It is an interface member, but it gives a stronger assertion. It ensures that the given key is a `StableNodeKey` but _also_ that it _belongs_ to this `NodeKeyManager`. This is because there's a scenario that I forgot about when we last talked: you might have a `StableNodeKey` from another app, or another session (or maybe you even have two Fluid containers in a single app, each with a shared tree, and you are mixing the StableNodeKeys from each together). In that case, you want to check that the `StableNodeKey` passed here is one of "your" `StableNodeKeys`, otherwise you can'd do anything with it.

I think it probably makes sense to just do both. Have a free function `isStableNodeKey(string)` and a method on `NodeKeyManager` called `hasKey(key: StableNodeKey)`. The implementation of the latter is simply `compressor.tryRecompress(key) !== undefined`.

Comment (daesunp) [reply packages/dds/tree/src/feature-libraries/node-key/nodeKeyManager.ts]: Removed method from the interface and created free function `isStableNodeKey`

File: packages/dds/tree/src/test/simple-tree/treeApi.spec.ts [hunk 11]
```diff
@@ -148,6 +148,18 @@ describe("treeApi", () => {
...
+				identifier: schema.identifier,
+			});
+			const config = new TreeConfiguration(schemaWithIdentifier, () => ({
+				identifier: "invalidUUID",
+			}));
+
+			const root = getView(config).root;
+
+			assert.equal(Tree.shortId(root), "invalidUUID");
+		});
```
Comment (noencke) [inline packages/dds/tree/src/test/simple-tree/treeApi.spec.ts:162]: I would recommend adding one more test for an identifier which is a valid UUID, but not a compressible UUID (i.e. not a stable key created by the compressor - just some UUID that passes `isStableNodeKey`)

**Review by noencke — APPROVED** [review]
